store/neo4j.go

```diff
@@ -0,0 +1,213 @@
+)
+
+type Store struct {
+	dbPool bolt.ClosableDriverPool
+}
+
+type neoArgMap map[string]interface{}
+
+type Storer interface {
+	Close(ctx context.Context) error
+	GetCodelist(hierarchy *models.Hierarchy) (string, error)
+	GetHierarchy(hierarchy *models.Hierarchy) (*models.Response, error)
+	GetCode(hierarchy *models.Hierarchy, code string) (*models.Response, error)
+}
+
+func New(dbUrl string) (Storer, error) {
+	pool, err := bolt.NewClosableDriverPool(dbUrl, 5)
+	if err != nil {
+		log.Error(err, nil)
+		return nil, err
+	}
+	return &Store{dbPool: pool}, nil
+}
+
+func (s Store) Close(ctx context.Context) error {
+	errChan := make(chan error)
+	go func() {
+		errChan <- s.dbPool.Close()
+	}()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (s Store) GetCodelist(hierarchy *models.Hierarchy) (string, error) {
+	neoStmt := fmt.Sprintf(existStmt, hierarchy.InstanceId, hierarchy.Dimension)
+	logData := log.Data{"statement": neoStmt}
+
+	log.Trace("executing exists query", logData)
+	conn, err := s.dbPool.OpenPool()
+	if err != nil {
+		return "", err
+	}
+	defer conn.Close()
+	rows, err := conn.QueryNeo(neoStmt, nil)
+	if err != nil {
+		return "", err
+	}
+	defer rows.Close()
+
+	data, _, err := rows.All()
+	if len(data) == 0 {
+		return "", nil
+	}
+	props := data[0][0].(graph.Node).Properties
+	return props["codeList"].(string), nil
+}
+
+func (s Store) GetHierarchy(hierarchy *models.Hierarchy) (*models.Response, error) {
+	neoStmt := fmt.Sprintf(getHierStmt, hierarchy.InstanceId, hierarchy.Dimension)
+	return s.QueryResponse(hierarchy, neoStmt, neoArgMap{})
+}
+
+func (s Store) GetCode(hierarchy *models.Hierarchy, code string) (res *models.Response, err error) {
+	// res = &models.Response{}
+	neoStmt := fmt.Sprintf(getCodeStmt, hierarchy.InstanceId, hierarchy.Dimension)
+	if res, err = s.QueryResponse(hierarchy, neoStmt, neoArgMap{"code": code}); err != nil {
+		return
+	}
+	res.Breadcrumbs, err = s.GetAncestry(hierarchy, code)
+	return
+}
+
+// QueryResponse performs DB query (neoStmt, neoArgs) returning Response (should be singular)
+func (s Store) QueryResponse(hierarchy *models.Hierarchy, neoStmt string, neoArgs neoArgMap) (res *models.Response, err error) {
+	logData := log.Data{"statement": neoStmt, "row_count": 0, "neo_args": neoArgs}
+
+	log.Trace("QueryResponse executing get query", logData)
+	conn, err := s.dbPool.OpenPool()
+	if err != nil {
+		return
+	}
+	defer conn.Close()
+	rows, err := conn.QueryNeo(neoStmt, neoArgs)
+	if err != nil {
+		return
+	}
+
+	res = &models.Response{}
+	countRows := 0
+	for row, meta, err := rows.NextNeo(); err == nil; row, meta, err = rows.NextNeo() {
+		countRows++
+		logData["row_count"] = countRows
+		logData["row"] = row
+		logData["meta"] = meta
+		log.Trace("QueryResponse db row", logData)
+		if countRows > 1 {
+			err = errors.New("QueryResponse: got more than one row from DB")
```

Comment:
The assignment to err has no effect. It assigns to the err variable in the for loop, which shadows the err declared outside of the for loop?